src/Features/CSharp/Portable/CodeRefactorings/UseRecursivePatterns/UseRecursivePatternsCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,518 @@
+        }
+
+        private static Func<SyntaxNode, SyntaxNode>? GetReplacementFunc(SyntaxNode? node, SemanticModel model)
+        {
+            return node switch
+            {
+                BinaryExpressionSyntax(LogicalAndExpression) logicalAnd
+                    => CombineLogicalAndOperands(logicalAnd, model),
+                CasePatternSwitchLabelSyntax { WhenClause: { } whenClause } switchLabel
+                    => CombineWhenClauseCondition(switchLabel.Pattern, whenClause.Condition, model),
+                SwitchExpressionArmSyntax { WhenClause: { } whenClause } switchArm
+                    => CombineWhenClauseCondition(switchArm.Pattern, whenClause.Condition, model),
+                WhenClauseSyntax { Parent: CasePatternSwitchLabelSyntax switchLabel } whenClause
+                    => CombineWhenClauseCondition(switchLabel.Pattern, whenClause.Condition, model),
+                WhenClauseSyntax { Parent: SwitchExpressionArmSyntax switchArm } whenClause
+                    => CombineWhenClauseCondition(switchArm.Pattern, whenClause.Condition, model),
+                _ => null
+            };
+        }
+
+        private static Func<SyntaxNode, SyntaxNode>? CombineLogicalAndOperands(BinaryExpressionSyntax logicalAnd, SemanticModel model)
+        {
+            if (TryDetermineReceiver(logicalAnd.Left, model) is not var (leftReceiver, leftTarget, leftFlipped) ||
+                TryDetermineReceiver(logicalAnd.Right, model) is not var (rightReceiver, rightTarget, rightFlipped))
+            {
+                return null;
+            }
+
+            // If we have an is-expression on the left, first we check if there is a variable designation that's been used on the right-hand-side,
+            // in which case, we'll convert and move the check inside the existing pattern, if possible.
+            // For instance, `e is C c && c.p == 0` is converted to `e is C { p: 0 } c`
+            if (leftTarget.Parent is IsPatternExpressionSyntax isPatternExpression &&
+                TryFindVariableDesignation(isPatternExpression.Pattern, rightReceiver, model) is var (containingPattern, rightNamesOpt))
+            {
+                Debug.Assert(leftTarget == isPatternExpression.Pattern);
+                Debug.Assert(leftReceiver == isPatternExpression.Expression);
+                return root =>
+                {
+                    var rightPattern = CreatePattern(rightReceiver, rightTarget, rightFlipped);
+                    var rewrittenPattern = RewriteContainingPattern(containingPattern, rightPattern, rightNamesOpt);
+                    var replacement = isPatternExpression.ReplaceNode(containingPattern, rewrittenPattern);
+                    return root.ReplaceNode(logicalAnd, AdjustBinaryExpressionOperands(logicalAnd, replacement));
+                };
+            }
+
+            if (TryGetCommonReceiver(leftReceiver, rightReceiver, model) is var (commonReceiverOpt, leftNames, rightNames))
```

Comment:
so it feels odd that hhis says: TryGetCommonReceiver... but it can return a null common receiver (i.e. commonReceiverOpt), but this is still considered success.  can you doc what's going on here?